### PR TITLE
BRITECH-150 Added GET /threads && /threads/{id}

### PR DESCRIPTION
**Pull request by thullSL**
The GetThread(id) function has a hacky exception catching bit. The ienumerable passed back from the scan function is doing some weird stuff that we can't figure out and that was the only way to get to work. @jebbinSCL can attest to this.

### CODE REVIEW

**Comment by thullSL** [comment]
We will attempt to find a more..."engineered" solution to the exception problem however this solution works enough for the demo

File: Comms/ThreadService.fs [hunk 1]
```diff
@@ -29,12 +29,18 @@ let getEntityItems (id : string) : seq<Correspondence> =
...
+    Log.Information("{S}", entity);
     { Id = entity.Id
       Items = entity.Id |> getEntityItems }
 
 let getThread (id : string) = 
     Log.Information("Contacting DynamoDB for thread with Id: {id}", id)
-    context.Scan<ThreadDetail>(new ScanCondition("Id", ScanOperator.Equal, id))
+    let threads = context.Scan<ThreadEntity>(new ScanCondition("Id", ScanOperator.Equal, id))
+    try 
+        Some( entityToThread( Seq.head threads))
```
Comment (csscottlogic) [inline Comms/ThreadService.fs]: Extraneous space

**Comment by csscottlogic** [comment]
Can we have a quick description of the "weird stuff"?

**Comment by thullSL** [comment]
I will attempt it. In `threadService.getThread(id)` it scans the db for the entry with id and returns a iEnumberable as the result. If we do any null or empty checks on the return value they are always positive but if we just take the head it works. So

```
 let result = 
         if Seq.isEmpty threads then None
         else Some(Seq.head threads)
```

Always gives `None` back or throws an error where just having

```
let result = Seq.head threads
```
Give the correct answer, unless the id is 404, in which case we get an exception thrown.

**Comment by ceejsmith** [comment]
This probably has something to do with deferred execution. In Linq in C#, doing something like DbContext.Set<T>.Where(item => item.Id == 1) would return an IQueryable<T>. In other words, nothing actually gets executed on the database. For that, you have to trigger execution with something like ToList() or Count().
Bear in mind that an IEnumerable<T> can be infinite.
I'm not sure what Seq.isEmpty actually does, but when there's deferred execution involved, it might well always return true.
Getting the first member of the sequence will trigger execution, so if you're expecting ID to be unique, head is probably the way to go.

**Comment by thullSL** [comment]
We have solved the problem by turning the sequence explicitly into an array triggering the execution before checking for empty.

**Comment by ceejsmith** [comment]
That sounds like a good solution

**Comment by jebbinSCL** [comment]
:shipit:

File: Comms/ThreadsController.fs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+
+type ThreadsController() = 
+    inherit ApiController()
+    member x.Get() = 
+        Log.Information("Received GET request for last contact")
+        x.Request.CreateResponse(getThreads())
+
+    member x.Get(id : string) = 
+        Log.Information("Received GET request for last contact")
```
Comment (lclarksl) [inline Comms/ThreadsController.fs]: Maybe include the thread id in the logging to distinguish it from the other GET request?

Comment (csscottlogic) [inline Comms/ThreadsController.fs]: +1. There's no point in logging the same message in different methods.

Comment (thullSL) [inline Comms/ThreadsController.fs]: Fixed, even corrected the words to say threads instead of last-contact. :ghost:

**Comment by lclarksl** [comment]
:shipit:

**Pull request merged**